homeassistant/components/screenlogic/data.py

```diff
@@ -0,0 +1,348 @@
+):
+    """Process template data."""
+    entity_registry = er.async_get(coordinator.hass)
+
+    def cleanup_excluded_entity(entity_key: str):
+        """Remove entity if it exists."""
+        assert coordinator.config_entry
+        unique_id = f"{coordinator.config_entry.unique_id}_{entity_key}"
+        if entity_id := entity_registry.async_get_entity_id(
+            platform_domain, SL_DOMAIN, unique_id
+        ):
+            _LOGGER.debug(
+                "Removing existing entity '%s' per data inclusion rule", entity_id
+            )
+            entity_registry.async_remove(entity_id)
+
+    gateway = coordinator.gateway
+
+    for device, device_groups in supported_devices.items():
+        if "*" in device_groups:
+            indexed_values = device_groups.pop("*")
+            for index in gateway.get_data(device):
+                device_groups[index] = indexed_values
+        for group, group_values in device_groups.items():
+            for value_key, value_params in group_values.items():
+                data_path = (device, group, value_key)
+
+                entity_key = generate_unique_id(device, group, value_key)
+
+                if (
+                    inclusion_rule := value_params.get(EntityParameter.INCLUDED)
+                    or DEVICE_INCLUSION_RULES.get(device)
+                ) is not None:
+                    assert isinstance(inclusion_rule, ScreenLogicRule)
+                    if not inclusion_rule.test(gateway, data_path):
+                        cleanup_excluded_entity(entity_key)
+                        continue
+
+                try:
+                    value_data = gateway.get_data(*data_path, strict=True)
+                except KeyError:
+                    _LOGGER.debug("Failed to find %s", data_path)
+                    continue
+
+                sub_code = value_params.get(
+                    EntityParameter.SUBSCRIPTION_CODE
+                ) or DEVICE_SUBSCRIPTION.get(device)
+                assert sub_code is None or isinstance(sub_code, int)
+
+                enabled = True
+                if (
+                    enabled_rule := value_params.get(EntityParameter.ENABLED)
+                ) is not None:
+                    assert isinstance(enabled_rule, ScreenLogicRule)
+                    enabled = enabled_rule.test(gateway, data_path)
+
+                base_kwargs = {
+                    "data_path": data_path,
+                    "key": entity_key,
+                    "entity_category": value_params.get(
+                        EntityParameter.ENTITY_CATEGORY, EntityCategory.DIAGNOSTIC
+                    ),
+                    "entity_registry_enabled_default": enabled,
+                    "name": value_data.get(ATTR.NAME),
+                }
+
+                yield base_kwargs, BaseScreenLogicEntityData(
+                    data_path, enabled, entity_key, sub_code, value_data, value_params
+                )
+
```

Comment:
This return type seems overly complex.  Can it return a single data structure instead. I think it would be cleaner to not have `data` know anything about entities or platforms as well.

Reply:
Thats definitely part of it. I added `This looks like Home Assistant entity logic` to the areas that stood out.

I was trying to come up with a way to split up this function so it didn't know anything about Home Assistant entities and returned the base data for constructing a tuple/list of EntityDescriptions.  I'm not 100% sure thats a good path to go down, but I figured the first step to getting this easier to understand is to break it up a bit.

Reply:
Yea, agree on breaking it up a bit. It is kinda everything tossed into one function.  That'll be my first step.